### PR TITLE
feat(util): implement P-Square algorithm for streaming percentile calcualtion

### PR DESCRIPTION
**Pull request by 0xdeafbeef**
```
p2_single_insert        time:   [17.227 ns 17.248 ns 17.270 ns]
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe

p2_single_value_read    time:   [1.3073 ns 1.3074 ns 1.3075 ns]
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

p2_amortized_insert     time:   [5.0904 µs 5.0938 µs 5.0983 µs]
Found 31 outliers among 100 measurements (31.00%)
  17 (17.00%) low severe
  1 (1.00%) low mild
  4 (4.00%) high mild
  9 (9.00%) high severe

p2_insert_and_read      time:   [17.966 ns 17.984 ns 18.003 ns]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe
```

### CODE REVIEW

**Comment by 0xdeafbeef** [comment]
image

File: core/src/blockchain_rpc/client.rs [hunk 1]
```diff
@@ -132,19 +145,36 @@ impl BlockchainRpcClient {
...
+        // This will make broadcast timeout to be adaptive based on the network conditions
         for validator in targets.as_ref() {
             let client = client.clone();
             let validator = validator.clone();
             let request = request.clone();
-            futures.push(JoinTask::new(async move {
-                client.send_to_validator(validator, request).await
+            let this = self.clone();
+            // we are not using `JoinTask` here because we want to measure the time taken by the broadcast
+            futures.push(tokio::spawn(async move {
```
Comment (Rexagon) [inline core/src/blockchain_rpc/client.rs:159]: Use `JoinTask` to cancel spawned futures when this future cancels

File: util/src/lib.rs [hunk 2]
```diff
@@ -7,6 +7,8 @@ pub mod io;
 pub mod progress_bar;
 pub mod serde_helpers;
 pub mod time;
+
+pub mod timeouts;
```
Comment (Rexagon) [inline util/src/lib.rs:11]: Just use the `time` module

File: util/src/timeouts/mod.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    pub fn new_with_config(
+        quantile: f64,
+        window_length: Duration,
+        max_windows: usize,
+        clock: C,
+    ) -> Result<Self, p2::Error> {
+        if !(0.0..=1.0).contains(&quantile) {
+            return Err(p2::Error::InvalidQuantile(quantile));
+        }
+
```
Comment (Rexagon) [inline util/src/timeouts/mod.rs:38]: Add assert `!window_length.is_zero()`

File: core/src/blockchain_rpc/client.rs [hunk 4]
```diff
@@ -132,19 +145,36 @@ impl BlockchainRpcClient {
...
+                    .append(start.elapsed().as_millis() as i64);
+                res
             }));
         }
 
-        tokio::time::timeout(self.inner.broadcast_timeout, async {
-            while let Some(res) = futures.next().await {
+        let timeout = self.broadcast_timeout();
+        tokio::time::timeout(timeout, async {
+            // inner task timeout won't happen because outer task timeout is always <= inner task timeout
```
Comment (Rexagon) [inline core/src/blockchain_rpc/client.rs:176]: Then why is the inner timeout even needed?

Comment (Rexagon) [reply core/src/blockchain_rpc/client.rs:176]: If there is are neighbours that are "dying" and are slow to respond, this method can spawn too many broadcasts. For that case you might need to track a number of running futures (just an atomic) and consume the remaining of futures after the timeout and cancel them while the number is greater than some threshold

File: util/src/lib.rs (hunk 2 above)
Comment (0xdeafbeef) [reply util/src/lib.rs]: resolved

File: util/src/timeouts/mod.rs (hunk 3 above)
Comment (0xdeafbeef) [reply util/src/timeouts/mod.rs]: resolved

**Review by Rexagon — APPROVED** [review]

**Review by MrWad3r — APPROVED** [review]
lgtm except possible "out of ranges" e.g. in adjust_heights fn. Discussed in PM.

**Pull request merged**